### ISSUE TITLE
Default configuration permits HTML/JavaScript injection

### ISSUE THREAD

**Issue by cbs228** [mide/minecraft-overviewer#117]
Per the Overviewer [documentation](https://docs.overviewer.org/en/latest/signs/#filter-functions), text retrieved from Minecraft signs must be escaped to prevent HTML and/or JavaScript injection attacks. The default [configuration](https://github.com/mide/minecraft-overviewer/blob/master/config/config.py) which ships with this project permits players to inject arbitrary HTML tags into the Leaflet page. The injected HTML will become part of the DOM when when the user interacts with the sign marker. On click, it will be rendered into the `leaflet-popup-pane` div.

Per the Minecraft wiki, each sign contains four text fields named [Text1..Text4](https://minecraft.fandom.com/wiki/Sign#Block_data). Each line is a `TAG_String`. Tags of this type can individually accommodate up to [2^16 bytes](https://minecraft.fandom.com/wiki/NBT_format#TAG_definition) of JSON-wrapped text. This is more than enough space for a program of great complexity.

The normal Minecraft client does not permit unfettered access to the `TAG_String`'s entire storage space: it places a very low limit on the length of each line. This, as well as the Docker image's default behavior of inserting `<br/>` tags between lines, makes it difficult to inject JavaScript via a normal Minecraft client. With that said, it is easy to gain access to the NBT tag's entire storage space via command blocks or Op commands.

The following proof-of-concept exploit, tested on Minecraft v1.16.5, requires Op-level access:

```
/setblock ~1 ~ ~ minecraft:acacia_sign replace
/data modify block ~1 ~ ~ Text1 set value '{"text":"<style onload=\\"alert(\'test\');\\"/>"}'
```

When rendered to the Overviewer Map, this sign will popup a JavaScript alert when clicked. 

The proof-of-concept exploit is not specific to v1.16.5 and can probably be adapted for many other versions of Minecraft. This particular set of commands requires Op permission, but I suspect that a modded client may be able to convince the server to store more sign text than normal. I have not tested this. A quick search suggests that the vanilla client limits sign line length based on [resource pack](https://bugs.mojang.com/browse/MC-89242). Resource packs are trivial to modify locally. I am not sure what limits the server itself enforces, if any.

To inject a script or other content onto an Overviewer page,

1. Overviewer must be configured to extract and publish Minecraft signs as markers. (This is a typical configuration.)
2. An authorized user account on the Minecraft server must create a sign that includes script or HTML content.
3. The victim must then explicitly interact with the generated sign marker. Unless otherwise mitigated by a Content Security Policy, any injected JavaScript will run in the Overviewer/Leaflet's script security context.

## Fix for the Docker Image (this project)

The upstream project [recommends](https://docs.overviewer.org/en/latest/signs/#filter-functions) using python's [`html.escape()`](https://docs.python.org/3/library/html.html#html.escape) method to escape characters of special significance to HTML. I recommend applying a similar fix here.

## Fix for the Upstream Project

The need to escape text stems from Overviewer's use of the Leaflet API,

[`overviewer_core/data/js_src/util.js`](https://github.com/overviewer/Minecraft-Overviewer/blob/03cb8eeaebe5052ebf60be7e86fcb0a7188caf18/overviewer_core/data/js_src/util.js#L418)
```js
if (marker_entry.createInfoWindow && db.text) {
    layerObj.bindPopup(db.text);
}
```

which eventually invokes the Leaflet method with `content = db.text`,

[`src/layer/DivOverlay.js`](https://github.com/Leaflet/Leaflet/blob/f10b44b3afcd079febe5219f84dc67c68b379b5e/src/layer/DivOverlay.js#L280)
```js
if (typeof content === 'string') {
    node.innerHTML = content; // <--- injection vulnerability
} else {
    while (node.hasChildNodes()) {
        node.removeChild(node.firstChild);
    }
    node.appendChild(content);
}
```

A safer alternative would be to provide a complete DOM node as input here, which is officially [documented](https://leafletjs.com/reference.html#layer-bindpopup) in Leaflet. A text node would work just fine:

```js
if (marker_entry.createInfoWindow && db.text) {
    layerObj.bindPopup(document.createTextNode(db.text));
}
```

Here, the popup will contain the literal text `db.text` without the need to escape anything. Preserving the original line-by-line sign layout could be accomplished with a [CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/white-space) `white-space: pre;` property.

EDIT: for proof-of-concept and proposed fix for upstream project.

**Comment by mide** [comment]
Thank you for the report @cbs228 - I have opened #118 to address the issue you've described. Let's move discussions pertaining to the fixes to that PR, while leaving all discussions pertaining to the root issue here _(proof-of-concepts, details, etc)_. 

I appreciate the responsible disclosure initially, it gave me the time to get a fix ready. Whenever you are ready, feel free to add any additional details / the proof-of-concept you initially sent.

**Comment by mide** [comment]
_Whoops, this was automatically closed by GitHub because of the commit message on 81ace3ad8362400283b2b4f0bc8f230a4d6096ad. Reopening._

**Comment by cbs228** [comment]
Edited for proof-of-concept. I think we can close this now (fixed in #118). Thanks for fixing this!
